internal/app/referenceserver/raw_response.go

```diff
@@ -39,12 +39,22 @@ type rawResponseKey struct{}
 // been stored, those response writer interactions take precedence and no
 // raw response can be sent.)
 func rawResponder(handler http.Handler, errPrinter internal.Printer) http.Handler {
+	errorWriter := connect.NewErrorWriter()
 	return http.HandlerFunc(func(respWriter http.ResponseWriter, req *http.Request) {
 		testCaseName := req.Header.Get("x-test-case-name")
+		// This is the only hard failure. Without it, we cannot provide feedback.
+		// All other checks below write to stderr to provide feedback.
 		if testCaseName == "" {
-			// This is the only hard failure. Without it, we cannot provide feedback.
-			// All other checks below write to stderr to provide feedback.
-			http.Error(respWriter, "missing x-test-case-name header", http.StatusBadRequest)
+			msg := "missing x-test-case-name header"
+			if errorWriter.IsSupported(req) {
+				invalidArg := connect.NewError(connect.CodeInvalidArgument, errors.New(msg))
+				err := errorWriter.Write(respWriter, req, invalidArg)
+				if err != nil {
+					http.Error(respWriter, err.Error(), http.StatusInternalServerError)
+				}
+			} else {
+				http.Error(respWriter, msg, http.StatusBadRequest)
+			}
```

Comment:
This part LGTM. If you want to pull this out into a separate PR, we can go ahead and get it merged. (Though not a big deal if you just want to leave it here -- if this isn't obviously valuable outside of this PR, then maybe no need to do the work to carve out separate PR for it...).